File: python/day1.py
import sys

lines = list(map(int, sys.stdin.readlines()))


def total_fuel(m):
    m = m // 3 - 2
    return m + total_fuel(m) if m > 0 else 0


# Part I
print(sum([m // 3 - 2 for m in lines]))

# Part II
print(sum([total_fuel(m) for m in lines]))

File: haskell/day1.hs
import Data.List

main :: IO ()
main = do
  content <- readFile "./inputs/day1.input"
  let input = map read (lines content)
  let part1 = sum (map fuel input)
  let part2 = sum (map totalFuel input)
  print part1
  print part2

fuel :: Int -> Int
fuel mass = mass `div` 3 - 2

totalFuel :: Int -> Int
totalFuel = sum
          . tail
          . takeWhile (> 0)
          . iterate fuel